Make NodeSamples editable

OliBomby opened the pull request:
This change allows editing and modding of the `IHasRepeats.NodeSamples`. Also `SamplePointPiece` has been upgraded to include a 'addition bank', achieving parity with stable.

Still can't edit the custom sample index for legacy samples though. Not sure if we want to include that in the editor.

This checks two boxes in https://github.com/ppy/osu/issues/12020
Closes #22187
Closes #20595

![osu Game Rulesets Osu Tests_y2uv5bNj0B](https://user-images.githubusercontent.com/17460441/236848255-1ffaa493-ac55-4c24-b7cd-f0c3103dbb6c.png)

OliBomby committed:
Remove dependency on SamplesBindable

OliBomby committed:
remove editor beatmap

OliBomby committed:
create NodeSamplePointPiece

OliBomby committed:
fix updating wrong samples

OliBomby committed:
Add node sample pieces to timeline blueprint

OliBomby committed:
Abbreviate common bank names on timeline

OliBomby committed:
fix assigned hitsounds dont have bank or volume

OliBomby committed:
allow editing additions in sample point piece

OliBomby committed:
cleanup code duplication

OliBomby committed:
fix crash on multiselect and node sample piece popup

OliBomby committed:
add addition bank

Speykious commented:
Given this new sample bank menu:
![image](https://github.com/ppy/osu/assets/13885008/ce3a8d5c-fdf0-40a6-afb4-fad26dfccb4d)

I guess this now needs to be a dropdown.
![image](https://github.com/ppy/osu/assets/13885008/c36ef3ef-807f-4004-bab4-5e7bf054b3e5)

Speykious commented:
The right click menu on a hitobject would need to show the two sample banks as well, right?
![image](https://github.com/ppy/osu/assets/13885008/b4aac503-bbbd-48c4-8284-df29f80641ca)

peppy commented:
@OliBomby hoping to get to this soon, but just to confirm forward direction (I think we may have discussed this so forgive me):

This is adding parity back, but the end goal will be to split the "addition bank" to be applied on a per-sample level, correct? This will have implications on the UI when we get to that point (basically the stuff in this PR will be replaced), so I'm curious if you already explored that direction and confirmed it's going to be a much larger scoped change before we go ahead with this smaller piece.

peppy committed:
Merge branch 'master' into edit-nodesample

peppy committed:
Adjust `CreateHitSampleInfo` to handle additions correctly, rather than implementing locally

peppy commented:
A few initial things:

Please check https://github.com/ppy/osu/pull/23443/commits/114f12a79056cc20951d06aec1ade073899d684a. I've moved the local logic of deciding which bank to use on creating new additions to the common method, since it felt awkward doing the same thing in multiple places when this method was intended to cover those cases.

---

Maybe a limitation of the implementation, but attempting to change the addition bank when there are no additions present will reset it back to the "normal" bank. Either this needs to be fixed, or the text box should be disabled when there are not yet any additions. I'd probably go with the latter as it feels like a better UX.


https://github.com/ppy/osu/assets/191335/368d9525-7302-479d-80fa-27d4fa29e0cb




---

There's no test coverage of the new behaviour. It'd be nice to have at least a bit.

peppy reviewed (changes requested):
Initial pass, see attached feedback.

OliBomby commented:
> @OliBomby hoping to get to this soon, but just to confirm forward direction (I think we may have discussed this so forgive me):
> 
> This is adding parity back, but the end goal will be to split the "addition bank" to be applied on a per-sample level, correct? This will have implications on the UI when we get to that point (basically the stuff in this PR will be replaced), so I'm curious if you already explored that direction and confirmed it's going to be a much larger scoped change before we go ahead with this smaller piece.

Yes, the intention is to eventually replace this UI so you can apply banks on a per-sample level. The reason I didn't go that direction yet is because the file encoding hasn't quite caught up yet, so saving banks on a per-sample level isn't possible with the legacy encoder.

The purpose of this PR is to bring parity with stable so people can edit hitsounds just like on stable (minus the custom samples).

peppy commented:
Cool, we're on the same page then.

Make sure to review my follow-up comment with some actionable feedback.

OliBomby commented:
> I guess this now needs to be a dropdown.

@Speykious I guess making it a dropdown is alright, since the purpose right now is only to bring parity with stable. Though eventually it will need to be a textbox again.

> The right click menu on a hitobject would need to show the two sample banks as well, right?

I don't think this bank context menu is necessary in the first place. You can just use the hitsound sample point thingy below the object on the timeline.

> Please check https://github.com/ppy/osu/commit/114f12a79056cc20951d06aec1ade073899d684a. I've moved the local logic of deciding which bank to use on creating new additions to the common method, since it felt awkward doing the same thing in multiple places when this method was intended to cover those cases.

@peppy Thank you for reviewing my PR! I agree with the changes, this is much cleaner.

> Maybe a limitation of the implementation, but attempting to change the addition bank when there are no additions present will reset it back to the "normal" bank. Either this needs to be fixed, or the text box should be disabled when there are not yet any additions. I'd probably go with the latter as it feels like a better UX.

Yeah it's impossible to add an additions bank when there are no additions to add the bank to. I agree with disabling the option in such case.

> There's no test coverage of the new behaviour. It'd be nice to have at least a bit.

I'll add some simple test cases to make sure that using these controls actually updates the hitsounds.

@peppy Should I invert the colours of the sample point pieces on sliders? so the pieces on node samples are pink and the piece for the main hitobject `Samples` (controlling the sliderbody hitsound) is purple. I feel like that would be more intuitive since for sliders and circles the pink sample point pieces would then have the same practical effect.

peppy commented:
> Should I invert the colours of the sample point pieces on sliders? 

Can you take a screenshot showing what you mean?

OliBomby commented:
> Can you take a screenshot showing what you mean?

![osu Game Rulesets Osu Tests_mgrcCdqpPP](https://github.com/ppy/osu/assets/17460441/bdf4369f-aba0-4e11-8ced-fc3c11d749d4)
^ Current implementation

![osu Game Rulesets Osu Tests_nmPXprgFrg](https://github.com/ppy/osu/assets/17460441/6de52095-512b-45fc-ab1f-7f1b682af9e7)
^ Suggested change

I'm aware that there are always two sample point pieces overlapping at the sliderhead. I'm not really sure what to do about that. You could make them narrower and put them next to eachother like in my follow-up PR, or merge them into one piece somehow, or just remove the piece for the sliderbody bank entirely.

OliBomby committed:
Invert colors

OliBomby commented:
I just realized https://github.com/ppy/osu/commit/114f12a79056cc20951d06aec1ade073899d684a broke some special logic in `SamplePointPiece`. It's actually supposed to prioritize inheriting the bank from the specific nodesample instead of the samples of the hitobject. If you only have `h.CreateHitSampleInfo(sampleName)` it will fetch the bank from the sliderbody hitsound when adding an addition to a slider repeat which is wrong.

OliBomby committed:
Fix regressed bank inheriting behaviour on node samples

OliBomby committed:
Make relevantObject and relevantSamples instance variables

OliBomby committed:
hide addition bank field when no additions active

OliBomby committed:
Add PopoverContainer to TimelineTestScene

OliBomby committed:
Add sample point piece test

OliBomby committed:
add nodesample piece test

peppy commented on osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:
```diff
@@ -144,54 +201,222 @@ protected override void LoadComplete()
+            private string? getCommonAdditionBank() => relevantSamples.Select(GetAdditionBankValue).Distinct().Count() == 1 ? GetAdditionBankValue(relevantSamples.First()) : null;
+            private int? getCommonVolume() => relevantSamples.Select(GetVolumeValue).Distinct().Count() == 1 ? GetVolumeValue(relevantSamples.First()) : null;
+
+            private void update(Action<HitObject, IList<HitSampleInfo>> updateAction)
```
This whole flow is very hard to follow. I'm hoping we can find a way to simplify this, at least not passing an `Action` inside as it's done here.

OliBomby replied:
the update method is meant to work as a little macro to reduce code duplication in all the specific update methods. Would it be clear enough if i rename some variables and add xmldoc to clarify how the flow works? I think it's quite intuitive to use once you know what it means.

peppy commented on osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:
```diff
@@ -144,54 +201,222 @@ protected override void LoadComplete()
+                    if (samples.Any(s => s.Name == sampleName))
+                        return;
+
+                    // First try inheriting the sample info from the node samples instead of the samples of the hitobject
```
I had a lot of trouble understanding how this changed anything until I realised that the `samples` list here is actually augmented from the `HitObject`'s samples.

peppy commented on osu.Game.Tests/Visual/Editing/TestSceneTimelineHitObjectBlueprint.cs:
```diff
@@ -111,5 +117,148 @@ public void TestDisallowRepeatsOnZeroDurationObjects()
+        }
+
+        [Test]
+        public void TestNodeSamplePointPiece()
```
These tests need to clear the popover at the start or the end, currently running them more than once results in a failure.

https://github.com/ppy/osu/assets/191335/1ee0af08-bcf8-4809-bb10-c0fffbdc482f

peppy commented on osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:
```diff
@@ -109,33 +151,48 @@ private void load()
+                    updateAdditionBank(val.NewValue);
+                    updateAdditionBankPlaceholderText();
+                });
+                additionBank.OnCommit += (_, _) => updateAdditionBankText();
```
Can logic around here be simplified like this?

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs b/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
index 37a6fa8a22..f9e04a8ca7 100644
--- a/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs
@@ -178,13 +178,12 @@ private void load()
                 // this ensures that committing empty text causes a revert to the previous value.
                 bank.OnCommit += (_, _) => bank.Current.Value = getCommonBank();
 
-                updateAdditionBankPlaceholderText();
                 updateAdditionBankText();
                 updateAdditionBankActivated();
                 additionBank.Current.BindValueChanged(val =>
                 {
                     updateAdditionBank(val.NewValue);
-                    updateAdditionBankPlaceholderText();
+                    updateAdditionBankText();
                 });
                 additionBank.OnCommit += (_, _) => updateAdditionBankText();
 
@@ -257,12 +256,6 @@ private void updateBankPlaceholderText()
                 bank.PlaceholderText = string.IsNullOrEmpty(commonBank) ? "(multiple)" : string.Empty;
             }
 
-            private void updateAdditionBankPlaceholderText()
-            {
-                string? commonAdditionBank = getCommonAdditionBank();
-                additionBank.PlaceholderText = string.IsNullOrEmpty(commonAdditionBank) ? "(multiple)" : string.Empty;
-            }
-
             private void updateAdditionBankActivated()
             {
                 bool anyAdditions = relevantSamples.Any(o => o.Any(s => s.Name != HitSampleInfo.HIT_NORMAL));
@@ -280,6 +273,7 @@ private void updateAdditionBankText()
                 if (string.IsNullOrEmpty(commonAdditionBank)) return;
 
                 additionBank.Current.Value = commonAdditionBank;
+                additionBank.PlaceholderText = string.IsNullOrEmpty(commonAdditionBank) ? "(multiple)" : string.Empty;
             }
 
             private void updateVolume(int? newVolume)
```

I have more to say about this code, but this would be a first step in the right direction.

OliBomby replied:
Updating the text in the `ValueChanged` event has a weird side-effect that it will never allow you to clear the textbox, because changes do not propagate when the textbox is empty. I'll find a way to simplify it in a different way

bdach replied:
Not much has simplified here. Speaking from personal experience (and I've made the same mistake myself), having multiple update methods like `updateAdditionBankText` and `updateAdditionBankVisual` and `updateAdditionBank` (with a bool) is not good news. Ideally you'd want just one `updateState` that gets everything correct.

Can you explain what the issues are here *before* I try and go in with a hammer to simplify this?

OliBomby replied:
I'll explain the issues as thoroughly as possible;

The update methods `updateBank`, `updateAdditionBank`, `updateVolume`, `addHitSample`, and `removeHitSample` must all only trigger when the user edits their respective value, because these are quite destructive in a sense that they synchronize the property across all selected hit objects.

Visuals must be updated before the bindings are created that update the hitobject properties on change, to prevent the hitobjects from being changed immediately when opening the popup.

Properties must be updated immediately on change, to allow the visuals in the timeline to reflect real-time changes.

Placeholder texts must be updated on change, to prevent it from incorrectly showing "(multiple)" after the user changes a property on multiple objects and clears the textbox.

The text of the textbox must not be updated on change, because otherwise the user will be unable to temporarily clear the textbox, because that is not a valid value.

bdach replied:
I think I managed to straighten this out by hoisting some null checks out of the set methods (see https://github.com/ppy/osu/pull/23443/commits/0efa028e0a82dc192303e2b99bead7ce1db66280). Please check whether this preserves all behaviours you wanted to, tests are passing as far as I could tell.

OliBomby replied:
Looks good. I did some small cleanup and fixed an unrelated bug i found while testing the behaviours.

OliBomby committed:
reset popover at the end of sample tests

OliBomby committed:
merge updateAdditionBankPlaceholderText and updateAdditionBankActivated

OliBomby committed:
added xmldoc and renamed GetSamples

OliBomby committed:
rename sample update logic and add xmldoc for clarity

OliBomby commented:
Any idea why the `TestNodeSamplePointPiece` test keeps failing in the CI but locally it passes without issue?

peppy commented:
It fails locally for me too. Maybe it's not for you because you're not running the other tests in the class in sequence? Sounds like something isn't being reset correctly between test methods.

![JetBrains Rider 2023-06-01 at 11 32 22](https://github.com/ppy/osu/assets/191335/00b95a6b-d001-47f3-926b-cb0e9f60d325)

OliBomby commented:
I somehow failed to notice `TestSceneHitObjectSamplePointAdjustments` exists. I should move the new tests to there.

peppy commented:
> I'm aware that there are always two sample point pieces overlapping at the sliderhead. I'm not really sure what to do about that. You could make them narrower and put them next to eachother like in my follow-up PR, or merge them into one piece somehow, or just remove the piece for the sliderbody bank entirely.

I'm not sure where the purple came from, but it seems weird. If you're wanting to differentiate, I think they should be different pink shades at most.

OliBomby committed:
Revert "Add PopoverContainer to TimelineTestScene"

This reverts commit 8acfe6b58b1de2de3466315fc74b000a428860c1.

OliBomby committed:
add new behaviour tests to TestSceneHitObjectSampleAdjustments

OliBomby committed:
revert add sample point piece tests

OliBomby committed:
Change purple to darker pink

OliBomby committed:
Change position of sliderbody hitsound piece for less overlap

OliBomby committed:
Merge branch 'master' into edit-nodesample

OliBomby committed:
fix merge conflict

OliBomby commented:
> I'm not sure where the purple came from, but it seems weird. If you're wanting to differentiate, I think they should be different pink shades at most.

I've changed it to dark pink for the sliderbody hitsound and normal pink for the node samples. Also changed the positions a bit so the node samples are always on the correct X position. I think it looks good now.

![osu Game Rulesets Osu Tests_42dLfyNCnf](https://github.com/ppy/osu/assets/17460441/43d1bcbf-0afd-4e0f-9bc7-c0ee44b55086)

peppy committed:
Merge branch 'master' into edit-nodesample

OliBomby commented:
BTW I addressed all the feedback so the PR is ready for merge

OliBomby committed:
Ensure populated node samples so new objects have unique node sample lists

OliBomby committed:
Move sliderbody hs to middle of first span

OliBomby committed:
Add keybind for bank setting

OliBomby committed:
dont focus volume, so keybinds immediately available

OliBomby committed:
fix no nodesample test case

OliBomby committed:
fix code quality

OliBomby committed:
fix focus test

Tais993 commented:
Quick bump, no reviews yet?

bdach commented:
@Tais993 work not on https://github.com/orgs/ppy/projects/13 is currently being heavily deprioritised. I hope you understand.

peppy commented:
@OliBomby wanna fix conflicts on this one just to keep it up-to-date?

OliBomby committed:
Merge remote-tracking branch 'upstream/master' into edit-nodesample

OliBomby commented:
> @OliBomby wanna fix conflicts on this one just to keep it up-to-date?

Sure

bdach committed:
Merge branch 'master' into edit-nodesample

bdach committed:
Use established constants

bdach commented:
@OliBomby I fixed conflicts here, please double check resolution.

bdach committed:
Fix obsoletion warning

OliBomby commented on osu.Game/Rulesets/Objects/HitObject.cs:
```diff
@@ -223,8 +223,16 @@ public IList<HitSampleInfo> CreateSlidingSamples()
         {
-            if (Samples.FirstOrDefault(s => s.Name == HitSampleInfo.HIT_NORMAL) is HitSampleInfo existingSample)
-                return existingSample.With(newName: sampleName);
+            // As per stable, all non-normal "addition" samples should use the same bank.
+            if (sampleName != HitSampleInfo.HIT_NORMAL)
+            {
+                if (Samples.FirstOrDefault(s => s.Name != HitSampleInfo.HIT_NORMAL) is HitSampleInfo existingAddition)
+                    return existingAddition.With(newName: sampleName);
+            }
```
This change was made by @peppy 

It seems to be a code duplication cleanup for logic that determines the correct bank to apply to a new hitsample. For example if you are making a new hitwhistle sample which is an 'addition', then this code tries to find an existing addition sample to take the bank name of.

In stable, all addition samples have the same bank, because you can only set one bank for all addition samples. This doesn't have to be true in lazer, but currently lazer behaves the same as stable for this in that you can only set one bank for all addition samples.

bdach replied:
Duplication from where though? If it was deduplication I'd expect the same chunk of code to be gone from somewhere else. But I haven't seen any of that in this diff.

OliBomby replied:
Read https://github.com/ppy/osu/pull/23443#issuecomment-1567761667 for the reasoning.
I think this PR introduced the need for the chunk of code that was duplicated and subsequently deduplicated, that's why you don't see it in the final diff.

bdach replied:
Well the thing is that I'm not confident that it's a safe change to make. Yes `CreateHitSampleInfo()` is frequently used in editor contexts, but it also is frequently used in *non*-editor context. Such as:

https://github.com/ppy/osu/blob/a8a2e543d51788b1dc05ab318b54b46af76d0be7/osu.Game.Rulesets.Osu/Objects/Spinner.cs#L94
https://github.com/ppy/osu/blob/a8a2e543d51788b1dc05ab318b54b46af76d0be7/osu.Game.Rulesets.Taiko/UI/DrumSampleTriggerSource.cs#L43-L52

So as far as I can tell this probably changes sample playback behaviour for some objects. So I'd like to know whether this was a change made specifically for the editor flows, or something that is supposed to be valid for everything everywhere.

@peppy mind providing context / commentary on this when able?

OliBomby replied:
I did some analysis:

It seems the change allows SpinnerBonusTick to receive an addition bank if the spinner has additions.
This is probably incorrect because spinnerbonus-max and spinnerspin seem to both always use the primary bank.

For taiko Hit and StrongableHitObject, this change allows clap and finish samples added by the Type and IsStrong bindables to use the addition bank.
I think this change is desirable in this case because the user is able to change the addition bank (since this PR), so probably expects all the whistle, clap, and finish samples to have the same addition bank. It would be weird if for example the user sets the clap sample to the normal bank, sets IsStrong to true, and finds a finish sample with the soft bank from the primary bank.

For DrumSampleTriggerSource, this change makes it possible to play the rim and centre samples on different banks if the most valid object has different primary and addition banks. It adds a little bit of mapper intention but maybe we don't want this.

I'm not familiar with taiko hitsounding, but it seems like taiko always plays a single sample per hit so having two different banks on the same object might be a silly idea. If that's the case then we should probably go back and always take the bank of the hitnormal in taiko.

bdach replied:
So I looked into this myself a bit further and well.....

Here's a test map that I did: [Moderat - Rusty Nails - Copy.zip](https://github.com/user-attachments/files/15817315/Moderat.-.Rusty.Nails.-.Copy.zip)

The thing to note here is that I had all of the sample files just be a speech synthesizer that says the name of the bank. As such:

- For the spinner thing, now I believe it to be an irrelevant concern because stable did not support banks for `spinnerbonus` at all. So the fact this even works is a lazer-only thing and not really advertised as supported anywhere, so I'm fine with leaving it as "undefined behaviour".
- For the taiko thing... well let's just say neither `master` nor this PR follow what stable does.
  - On stable, I can't get the drum to produce anything audible other than "normal". Which means stable does not appear to care about sample bank on individual objects *at all*. The only thing that appears to have an effect there is the thing "default sample settings" (it'll be all normal, all soft, or all drum, always).
  - On `master`, every object with additions will follow its "primary" / `hitnormal` bank,
  - while on this branch, every object with additions will follow its "addition" bank.

I'm not sure what to think of the taiko thing. I'll go ask the taiko NAT maybe for some advice because I'm stupid and know nothing about mapping, but I'm not sure this needs to be a blocking concern for this PR.

With that said I'm going to just probably be looking to get this in and address any issues later. @ppy/team-client any objections to that?

peppy replied:
> I'm not sure what to think of the taiko thing.

As far as I've experienced, taiko players they expect to hear a taiko drum sound and hitsounds either take a backseat being "whatever" or (as we did in lazer) are more present to the point of annoying them.

> With that said I'm going to just probably be looking to get this in and address any issues later. @ppy/team-client any objections to that?

Sounds fine.

OliBomby committed:
Fix wasteful recreating of container

OliBomby committed:
Merge branch 'edit-nodesample' of https://github.com/OliBomby/osu into edit-nodesample

OliBomby committed:
undo useless change

OliBomby committed:
Add some in-depth xmldoc to GetSamples

OliBomby commented on osu.Game/Screens/Edit/Compose/Components/Timeline/SamplePointPiece.cs:
```diff
@@ -50,28 +55,54 @@ protected override bool OnClick(ClickEvent e)
         }
 
-        public Popover GetPopover() => new SampleEditPopover(HitObject);
+        protected virtual IList<HitSampleInfo> GetSamples() => HitObject.Samples;
```
It seems I addressed it by adding xmldoc to `SampleEditPopoer.GetRelevantSamples()`, but now I've added it to `GetSamples` too

OliBomby commented on osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs:
```diff
@@ -236,6 +244,27 @@ private void updateRepeats(IHasRepeats repeats)
+            sampleComponents.Add(nodeSamplesContainer = new Container
+            {
+                RelativeSizeAxes = Axes.Both,
+            });
```
Just clearing contents seems to work fine

bdach committed:
Restructure popover updates to be more centralised

bdach committed:
Use slightly lighter shade of pink for alternative colour

bdach commented:
> I've changed it to dark pink for the sliderbody hitsound and normal pink for the node samples.

I felt that the contrast with this between the dark pink bg and the black text was pretty dreadful so I lightened it up a touch in https://github.com/ppy/osu/pull/23443/commits/7d5dc750e53938d60fa6cffedc1879d32c546118:

![1718014272](https://github.com/ppy/osu/assets/20418176/0744243e-252e-468e-857e-ff806099ea12)

The only remaining concern to me with this is [this review thread](https://github.com/ppy/osu/pull/23443#discussion_r1629371984). If we can revert this change / localise it to the editor, or justify its correctness in all contexts _somehow_, then I think this is good to go.

OliBomby commented:
I like the change to the contrast of the alternative colour.

> The only remaining concern to me with this is https://github.com/ppy/osu/pull/23443#discussion_r1629371984. If we can revert this change / localise it to the editor, or justify its correctness in all contexts somehow, then I think this is good to go.

I'll look into it to see what this change exactly means for all the usages of this method.

OliBomby committed:
remove code already covered by updatePrimaryBankState

OliBomby committed:
Fixed samples without additions contributing to common addition bank while not having an editable addition bank

smoogipoo commented on osu.Game.Rulesets.Catch/Objects/JuiceStream.cs:
```diff
@@ -76,6 +76,8 @@ protected override void CreateNestedHitObjects(CancellationToken cancellationTok
         {
             base.CreateNestedHitObjects(cancellationToken);
 
+            this.PopulateNodeSamples();
```
So is every single implementation of `CreateNestedHitObjects` going to have to call this?

bdach replied:
Every `IHasRepeats` would need to I suppose. Not entirely sure how to fix that without introducing eldritch abstract entities.

peppy replied:
@smoogipoo do you have strong opinions on this?

smoogipoo replied:
not at the moment, no

bdach commented:
One other thing I didn't really consider (which was pointed out to me in a [reddit discussion](https://www.reddit.com/r/osugame/comments/1dga6m5/comment/l8s0gms/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button)) is that this UI does not work with buzz sliders:

![1718605577](https://github.com/ppy/osu/assets/20418176/d0b5eabd-5aa7-4183-9ea9-d66cee0c7b3e)
![1718605587](https://github.com/ppy/osu/assets/20418176/75ba7bb9-6b92-4364-b937-edea850af869)

@OliBomby thoughts on how we can remedy this?

OliBomby commented:
I can think of a few workarounds but the UX leaves a lot to be desired. As a workaround you can zoom in really far in the timeline to create more space so the sample points dont overlap anymore.
You could also multiselect another slider and edit the body hitsounds of the other to hit the buzz slider body at the same time.

The information density of buzz sliders is pretty hard to handle on the timeline. Maybe a fix could be to merge the sample point pieces into a single expanding box whenever they overlap too much. Clicking it would then expend it down vertically and show every node to be edited. This still would work for super long buzz sliders though.

bdach commented:
The more these sorts of issues pop up, the more I think we need a new dedicated screen for hitsounding. A few of the ideas you proposed popped up in my head too but I'm not super sure they are going to work well in practice.

OliBomby commented:
> The more these sorts of issues pop up, the more I think we need a new dedicated screen for hitsounding. A few of the ideas you proposed popped up in my head too but I'm not super sure they are going to work well in practice.

I want to add a dedicated screen for hitsounding too, alongside being able to do basic hitsounds in the compose screen. This is because [some people prefer hitsounding while mapping](https://x.com/OliBomby/status/1637856574848000001). 

A dedicated screen for hitsounding will definitely make hitsounding a lot better of an experience for anything that isn't just adding simple whistle, finish, claps.
